src/channel_internal.c

```diff
@@ -279,148 +353,166 @@ static int dequeue_operation(THANDLE(CHANNEL_INTERNAL) channel_internal, THANDLE
-                LogError("Failure in dequeue_operation(channel_internal=%p, out_op_pull=%p, correlation_id=%" PRI_RC_STRING ", pull_callback=%p, pull_context=%p, NULL, NULL, NULL, NULL)", channel_internal, out_op_pull, RC_STRING_VALUE_OR_NULL(correlation_id), pull_callback, pull_context);
-                result = CHANNEL_RESULT_ERROR;
+                if (enqueue_operation(channel_internal, out_op_pull, correlation_id, on_data_available_cb, pull_context, NULL, NULL, NULL, NULL) != 0)
+                {
+                    /*Codes_SRS_CHANNEL_INTERNAL_43_023: [ If there are any failures, channel_internal_pull shall fail and return CHANNEL_RESULT_ERROR. ]*/
+                    LogError("Failure in enqueue_operation(channel_internal=%p, out_op_pull=%p, correlation_id=%" PRI_RC_STRING ", on_data_available_cb=%p, pull_context=%p, NULL, NULL, NULL, NULL)", channel_internal, out_op_pull, RC_STRING_VALUE_OR_NULL(correlation_id), on_data_available_cb, pull_context);
+                    result = CHANNEL_RESULT_ERROR;
+                }
+                else
+                {
+                    /*Codes_SRS_CHANNEL_INTERNAL_43_011: [ channel_internal_pull shall succeeds and return CHANNEL_RESULT_OK. ]*/
+                    result = CHANNEL_RESULT_OK;
+                }
             }
+            /*Codes_SRS_CHANNEL_INTERNAL_43_108: [ If the first operation in the list of pending operations contains a non-NULL on_data_consumed_cb: ]*/
             else
             {
-                /* Codes_SRS_CHANNEL_INTERNAL_43_011: [ channel_internal_pull shall succeeds and return CHANNEL_RESULT_OK. ]*/
-                result = CHANNEL_RESULT_OK;
+                if (dequeue_operation(channel_internal, out_op_pull, correlation_id, on_data_available_cb, pull_context, NULL, NULL, NULL, NULL) != 0)
+                {
+                    /*Codes_SRS_CHANNEL_INTERNAL_43_023: [ If there are any failures, channel_internal_pull shall fail and return CHANNEL_RESULT_ERROR. ]*/
+                    LogError("Failure in dequeue_operation(channel_internal=%p, out_op_pull=%p, correlation_id=%" PRI_RC_STRING ", on_data_available_cb=%p, pull_context=%p, NULL, NULL, NULL, NULL)", channel_internal, out_op_pull, RC_STRING_VALUE_OR_NULL(correlation_id), on_data_available_cb, pull_context);
+                    result = CHANNEL_RESULT_ERROR;
+                }
+                else
+                {
+                    /* Codes_SRS_CHANNEL_INTERNAL_43_011: [ channel_internal_pull shall succeeds and return CHANNEL_RESULT_OK. ]*/
+                    result = CHANNEL_RESULT_OK;
+                }
             }
+            /*Codes_SRS_CHANNEL_INTERNAL_43_115: [ channel_internal_pull shall call srw_lock_release_exclusive. ]*/
+            srw_lock_release_exclusive(channel_internal_ptr->lock);
         }
-
+        result == CHANNEL_RESULT_OK ? (void)0 : sm_exec_end(channel_internal_ptr->sm);
```

Comment:
Every time I read a line of code that looks like this, I die a little bit inside.

Reply:
Open to suggestions